File: src/modules/memory.ts
import { invoke } from "@tauri-apps/api/core";

export interface ProfileEntry {
  key: string;
  value: string;
}

export interface MemoryMessage {
  session_id: string;
  role: string;
  content: string;
  created_at: number;
}

let sessionId: string | null = null;

function generateId(): string {
  return Date.now().toString(36) + Math.random().toString(36).slice(2, 8);
}

export function getSessionId(): string {
  if (!sessionId) {
    sessionId = generateId();
  }
  return sessionId;
}

export async function startSession(): Promise<void> {
  sessionId = generateId();
  await invoke("memory_start_session", { sessionId });
  console.log(
    `%c[session]%c ${sessionId}`,
    "color: #f0c040; font-weight: bold",
    "color: #8ab4f8",
  );
}

export async function endSession(): Promise<void> {
  if (!sessionId) return;
  await invoke("memory_end_session");
  console.log(
    `%c[session]%c ${sessionId} %cended`,
    "color: #f0c040; font-weight: bold",
    "color: #8ab4f8",
    "color: #e04444",
  );
}

export async function saveMessage(
  role: "user" | "assistant",
  content: string,
): Promise<void> {
  await invoke("memory_save_message", {
    sessionId: getSessionId(),
    role,
    content,
  });
}

export async function searchMessages(
  query: string,
  limit = 20,
  since?: number,
): Promise<MemoryMessage[]> {
  return invoke("memory_search", { query, limit, since });
}

export interface SessionInfo {
  id: string;
  started_at: number;
  ended_at: number | null;
  msg_count: number;
}

export interface SessionMessage {
  role: string;
  content: string;
  created_at: number;
}

export async function listSessions(limit = 5): Promise<SessionInfo[]> {
  return invoke("memory_list_sessions", { limit });
}

export async function loadSession(
  sessionId: string,
): Promise<SessionMessage[]> {
  return invoke("memory_load_session", { sessionId });
}

export async function getProfile(): Promise<ProfileEntry[]> {
  return invoke("memory_get_profile");
}

export async function upsertProfile(key: string, value: string): Promise<void> {
  await invoke("memory_upsert_profile", { key, value });
}

File: shell/src/commands/memory.rs
use crate::db::Database;
use serde::Serialize;
use tauri::State;

#[derive(Serialize)]
pub struct ProfileEntry {
    key: String,
    value: String,
}

#[tauri::command]
pub fn memory_start_session(db: State<'_, Database>, session_id: String) -> Result<(), String> {
    db.start_session(&session_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn memory_end_session(db: State<'_, Database>) -> Result<(), String> {
    db.end_current_session().map_err(|e| e.to_string())
}

#[tauri::command]
pub fn memory_save_message(
    db: State<'_, Database>,
    session_id: String,
    role: String,
    content: String,
) -> Result<(), String> {
    db.save_message(&session_id, &role, &content)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn memory_list_sessions(
    db: State<'_, Database>,
    limit: u32,
) -> Result<Vec<serde_json::Value>, String> {
    let results = db.list_sessions(limit).map_err(|e| e.to_string())?;
    let json = results
        .into_iter()
        .map(|(id, started_at, ended_at, msg_count)| {
            serde_json::json!({
                "id": id,
                "started_at": started_at,
                "ended_at": ended_at,
                "msg_count": msg_count,
            })
        })
        .collect();
    Ok(json)
}

#[tauri::command]
pub fn memory_load_session(
    db: State<'_, Database>,
    session_id: String,
) -> Result<Vec<serde_json::Value>, String> {
    let results = db
        .load_session_messages(&session_id)
        .map_err(|e| e.to_string())?;
    let json = results
        .into_iter()
        .map(|(role, content, created_at)| {
            serde_json::json!({
                "role": role,
                "content": content,
                "created_at": created_at,
            })
        })
        .collect();
    Ok(json)
}

#[tauri::command]
pub fn memory_search(
    db: State<'_, Database>,
    query: String,
    limit: u32,
) -> Result<Vec<serde_json::Value>, String> {
    let results = db
        .search_messages(&query, limit)
        .map_err(|e| e.to_string())?;
    let json = results
        .into_iter()
        .map(|(sid, role, content, created_at)| {
            serde_json::json!({
                "session_id": sid,
                "role": role,
                "content": content,
                "created_at": created_at,
            })
        })
        .collect();
    Ok(json)
}

#[tauri::command]
pub fn memory_get_profile(db: State<'_, Database>) -> Result<Vec<ProfileEntry>, String> {
    let rows = db.get_profile().map_err(|e| e.to_string())?;
    Ok(rows
        .into_iter()
        .map(|(key, value)| ProfileEntry { key, value })
        .collect())
}

#[tauri::command]
pub fn memory_upsert_profile(
    db: State<'_, Database>,
    key: String,
    value: String,
) -> Result<(), String> {
    db.upsert_profile(&key, &value).map_err(|e| e.to_string())
}
